README.md

```diff
@@ -2,18 +2,126 @@
+       (conform-or-log ::coercer/location)))
+```
+
+The return value of requests will be converted to camelCase json and returned to the client. If the return value looks like `{:error ...}`, it is assumed to indicate an error response, and the `...` part will be set as the `error` of a [JSON-RPC error object](https://www.jsonrpc.org/specification#error_object). It is up to you to conform the `...` object (by giving it a `code`, `message`, and `data`.) Otherwise, the entire return value will be set as the `result` of a [JSON-RPC response object](https://www.jsonrpc.org/specification#response_object). (Message ids are handled internally by lsp4clj.)
+
+### Send messages
+
+Servers also initiate their own requests and notifications to a client. To send a notification, call `lsp4clj.server/send-notification`.
+
+```clojure
+(->> {:message message
+      :type type
+      :extra extra}
+     (conform-or-log ::coercer/show-message)
+     (lsp4clj.server/send-notification server "window/showMessage"))
+```
+
+Sending a request is similar, with `lsp4clj.server/send-request`. This method returns a request object which may be dereffed to get the client's response. Most of the time you will want to call `lsp4clj.server/deref-or-cancel`, which will send a `$/cancelRequest` to the client if a timeout is reached before the client responds.
+
+```clojure
+(let [request (->> {:edit edit}
+                   (conform-or-log ::coercer/workspace-edit-params)
+                   (lsp4clj.server/send-request server "workspace/applyEdit"))
+      response (lsp4clj.server/deref-or-cancel request 10e3 ::timeout)]
+  (if (= ::timeout response)
+    (logger/error "No reponse from client after 10 seconds.")
+    response))
+```
+
+Otherwise, the request object presents the same interface as `future`. Responds to `future-cancel` (which sends `$/cancelRequest`), `realized?`, `future?`, `future-done?` and `future-cancelled?`.
+
+If the request is cancelled, future invocations of `deref` will return `:lsp4clj.server/cancelled`.
+
+Sends `$/cancelRequest` only once, though `lsp4clj.server/deref-or-cancel` or `future-cancel` can be called multiple times.
+
+### Start and stop a server
+
+The last step is to start the server you created earlier. Use `lsp4clj.server/start`. This method accepts two arguments, the server and a "context".
+
+The context should be `associative?`. Whatever you provide in the context will be passed as the second argument to the notification and request `defmethod`s you defined earlier. This is a convenient way to make components of your system available to those methods without definining global constants. Often the context will include the server itself so that you can initiate outbound requests and notifications in reaction to inbound messages. lsp4clj reserves the right to add its own data to the context, using keys namespaced with `:lsp4clj.server/...`.
+
+```clojure
+(lsp4clj.server/start server {:server server, :logger logger})
```

Comment:
Maybe use another example here than passing the server itself to make it clear for people who don't know the components concept?
maybe pass logger and some other map like `:my-client-settings` or something like that

Reply:
Just a nit though

Reply:
Changed